src/rules/component-logic-loc.js

```diff
@@ -0,0 +1,111 @@
+            minimum: 0,
+            type: 'integer',
+          },
+          skipBlankLines: {
+            type: 'boolean',
+          },
+          skipComments: {
+            type: 'boolean',
+          },
+        },
+        type: 'object',
+      },
+    ],
+  },
+};
+
+const create = (context) => ({
+  // counts the number of lines starting from the first line of the component logic
+  // to the return statement
+  ReturnStatement: (node) => {
+    const maxLines = context.options[0]?.max ?? 100;
+    const skipBlankLines = context.options[0]?.skipBlankLines ?? false;
+    const skipComments = context.options[0]?.skipComments ?? false;
+    if (node.argument?.type !== 'JSXElement') {
+      return;
+    }
+    // find the function declaration node
+    let { parent } = node;
+    while (
+      parent?.type != null &&
+      parent.type !== 'FunctionDeclaration' &&
+      parent.type !== 'ArrowFunctionExpression'
+    ) {
+      ({ parent } = parent);
+    }
+    // find the function body node
+    let { body } = parent;
+    while (body.type !== 'BlockStatement') {
+      ({ body } = body);
+    }
+    // count the number of lines
+    let lines = node.loc.start.line - body.loc.start.line;
+
+    // exclude blank lines
+    if (skipBlankLines) {
+      const linesOfCode = context
+        .getSourceCode()
+        .lines.slice(body.loc.start.line - 1, node.loc.start.line - 1);
+      const blankLines = linesOfCode.filter((line) => line.trim() === '');
+      lines -= blankLines.length;
+    }
+
+    // exclude comments
+    if (skipComments) {
+      body.comments.forEach((comment) => {
+        if (
+          comment.loc.start.line >= body.loc.start.line &&
+          comment.loc.start.line <= node.loc.start.line
+        ) {
+          lines -= comment.loc.end.line - comment.loc.start.line + 1;
+        }
+      });
+    }
+
+    // exclude early return statements
+    // traverse the body of the function and find all return statements
+    // that are before the return statement of the component logic
+    const returnStatements = [];
+    const queue = [body];
+    while (queue.length > 0) {
+      const current = queue.shift();
+      if (current.type === 'ReturnStatement') {
+        returnStatements.push(current);
+      }
+      if (current.body != null) {
+        queue.push(...current.body);
+      }
+      if (current.consequent != null) {
+        queue.push(current.consequent);
+      }
+      if (current.alternate != null) {
+        queue.push(current.alternate);
+      }
+    }
+    returnStatements.forEach((returnStatement) => {
+      if (returnStatement.loc.start.line < node.loc.start.line) {
+        lines -= returnStatement.loc.end.line - returnStatement.loc.start.line + 1;
+      }
+    });
+
+    if (lines > 100) {
+      context.report({
+        message: `Component logic is ${lines} lines (${body.loc.start.line} - ${node.loc.start.line}), which is greater than the maximum of ${maxLines} lines\n`,
```

Comment:
Neat! 
Should we also have a cap on the LOC in the entire file?

Reply:
I'm not sure that's a good idea because of JSX. That seems like it would lead to unnecessary splitting up of JSX. If it's just a lot of rendering with no logic, I think having a long section of html is fine.

We could set a very generous limit, but that would diminish the effectiveness of such a rule.

Reply:
It would encourage thinking in terms of smaller components (which in turn makes it easier to add stories, tests, etc) and make the code more readable. Anyway, we can revisit it.

Reply:
That's an existing setting we can turn on whenever we want https://eslint.org/docs/latest/rules/max-lines